### PR TITLE
Add the Training Module

### PR DESCRIPTION
**Pull request by Seongoo**
### Summary

This PR adds support for training mode in the Physical AI Server. Key changes include:

-  A new Training Page in the Web UI

- Related UI components to support training interactions

### CODE REVIEW

**Review by ola31 — APPROVED** [review]
LGTM

File: physical_ai_server/physical_ai_server/physical_ai_server.py [hunk 1]
```diff
@@ -66,6 +73,19 @@ def __init__(self):
         self.robot_type_list = self.get_robot_type_list()
         self.start_recording_time: float = 0.0
 
+        self.training_thread = None
+        self.save_root_path = ''
+        self.is_training = False
+        self.training_status_timer = None
+        self.training_status_pub = self.create_publisher(
```
Comment (DongyunRobotis) [inline physical_ai_server/physical_ai_server/physical_ai_server.py]: The Publisher should be placed in the Communicator, or the structure could break.

File: physical_ai_server/physical_ai_server/physical_ai_server.py [hunk 2]
```diff
@@ -66,6 +73,19 @@ def __init__(self):
...
+            TrainingStatus,
+            '/training/status',
+            self.PUB_QOS_SIZE
+        )
+        self.WEIGHT_SAVE_ROOT_PATH = (
```
Comment (DongyunRobotis) [inline physical_ai_server/physical_ai_server/physical_ai_server.py]: Can't the WEIGHT_SAVE_ROOT_PATH be placed in the Training Manager?

File: physical_ai_server/physical_ai_server/physical_ai_server.py [hunk 3]
```diff
@@ -171,6 +201,25 @@ def init_ros_params(self, robot_type):
         self.get_logger().info(
             f'ROS parameters initialized successfully for robot type: {robot_type}')
 
+    def publish_training_status(self):
```
Comment (DongyunRobotis) [inline physical_ai_server/physical_ai_server/physical_ai_server.py]: The Publisher should be placed in the Communicator, or the structure could break.

File: physical_ai_server/physical_ai_server/physical_ai_server.py [hunk 4]
```diff
@@ -384,6 +433,73 @@ def _inference_timer_callback(self):
             self.timer_manager.stop(timer_name=self.operation_mode)
             return
 
+    def user_training_interaction_callback(self, request, response):
+        try:
+            if request.command == SendTrainingCommand.Request.START:
+                self.training_status_timer = self.create_timer(
+                    self.DEFAULT_TOPIC_TIMEOUT / 20,
```
Comment (DongyunRobotis) [inline physical_ai_server/physical_ai_server/physical_ai_server.py]: Why was Default_TOPIC_TIMEOUT divided by 20?
This constant is intended for checking the Topic's timeout. Please define the Timer Frequency as a constant and use that instead.

File: physical_ai_server/physical_ai_server/physical_ai_server.py [hunk 5]
```diff
@@ -504,6 +626,128 @@ def get_policy_list_callback(self, request, response):
         response.policy_list = policy_list
         return response
 
+    def get_available_list_callback(self, request, response):
+        try:
+            policy_list = [
```
Comment (DongyunRobotis) [inline physical_ai_server/physical_ai_server/physical_ai_server.py]: I think it's right for policy_list and device_list to be managed by the TrainingManager. It would be better if we could call and use them from the TrainingManager.

**Review by DongyunRobotis — CHANGES_REQUESTED** [review]

File: physical_ai_server/physical_ai_server/physical_ai_server.py [hunk 6]
```diff
@@ -504,6 +626,128 @@ def get_policy_list_callback(self, request, response):
...
+                'pi0',
+                'pi0fast',
+            ]
+
+            device_list = [
+                'cuda',
+                'cpu',
+            ]
+
+            missing_items = []
```
Comment (DongyunRobotis) [inline physical_ai_server/physical_ai_server/physical_ai_server.py]: I think this feature is unnecessary. Please remove it.

File: physical_ai_server/physical_ai_server/physical_ai_server.py (hunk 2 above)
Comment (DongyunRobotis) [reply physical_ai_server/physical_ai_server/physical_ai_server.py]: Alternatively, it would be better to place it in the section where constants are declared, like this:

DEFAULT_SAVE_ROOT_PATH = Path.home() / '.cache/huggingface/lerobot'
DEFAULT_TOPIC_TIMEOUT = 5.0  # seconds

File: physical_ai_server/physical_ai_server/physical_ai_server.py (hunk 5 above)
Comment (Seongoo) [reply physical_ai_server/physical_ai_server/physical_ai_server.py]: Thanks for pointing this out! Here is the corrected one like you suggested.

https://github.com/ROBOTIS-GIT/physical_ai_tools/commit/1a7d90eac5c1eee59ad99e190f805370685edad2

File: physical_ai_server/physical_ai_server/physical_ai_server.py (hunk 4 above)
Comment (Seongoo) [reply physical_ai_server/physical_ai_server/physical_ai_server.py]: This was just temporal. I will rename it!
https://github.com/ROBOTIS-GIT/physical_ai_tools/commit/d7d920e69f6616ddf22d80f51305f7ab0fe3ed4f

File: physical_ai_server/physical_ai_server/physical_ai_server.py (hunk 2 above)
Comment (Seongoo) [reply physical_ai_server/physical_ai_server/physical_ai_server.py]: Sounds better! 
https://github.com/ROBOTIS-GIT/physical_ai_tools/commit/a11238fc361d48fa2a23666981023049d26d1ab1

File: physical_ai_server/physical_ai_server/physical_ai_server.py (hunk 3 above)
Comment (Seongoo) [reply physical_ai_server/physical_ai_server/physical_ai_server.py]: How about it? :)

https://github.com/ROBOTIS-GIT/physical_ai_tools/commit/300fd774daadf145523c3ae5aea43e99e246589b

File: physical_ai_server/physical_ai_server/physical_ai_server.py [hunk 7]
```diff
@@ -66,6 +74,11 @@ def __init__(self):
         self.robot_type_list = self.get_robot_type_list()
         self.start_recording_time: float = 0.0
 
+        self.training_thread = None
+        self.save_root_path = ''
```
Comment (DongyunRobotis) [inline physical_ai_server/physical_ai_server/physical_ai_server.py]: It seems save_root_path is not being used. It's only being used as a local variable when passed into TrainingManager.

File: physical_ai_server/physical_ai_server/physical_ai_server.py [hunk 8]
```diff
@@ -384,6 +426,88 @@ def _inference_timer_callback(self):
             self.timer_manager.stop(timer_name=self.operation_mode)
             return
 
+    def user_training_interaction_callback(self, request, response):
+        try:
+            if request.command == SendTrainingCommand.Request.START:
+                self.training_manager = TrainingManager()
+                self.training_status_timer = self.create_timer(
```
Comment (DongyunRobotis) [inline physical_ai_server/physical_ai_server/physical_ai_server.py]: For timers, they're managed together by a class called TimerManager. Please modify this, referring to how TimerManager is used in other code.

**Review by DongyunRobotis — CHANGES_REQUESTED** [review]

File: physical_ai_server/physical_ai_server/physical_ai_server.py (hunk 6 above)
Comment (Seongoo) [reply physical_ai_server/physical_ai_server/physical_ai_server.py]: I guess this is outdated.

File: physical_ai_server/physical_ai_server/physical_ai_server.py (hunk 7 above)
Comment (Seongoo) [reply physical_ai_server/physical_ai_server/physical_ai_server.py]: Done!

https://github.com/ROBOTIS-GIT/physical_ai_tools/commit/d229b0fd941c177c5d8f5643f02680a5efb68a5d

File: physical_ai_server/physical_ai_server/physical_ai_server.py (hunk 8 above)
Comment (Seongoo) [reply physical_ai_server/physical_ai_server/physical_ai_server.py]: Done!
https://github.com/ROBOTIS-GIT/physical_ai_tools/pull/40/commits/ddfdae81ae66c80a7ee8f042c6555d012cd1564b

File: physical_ai_server/physical_ai_server/physical_ai_server.py (hunk 3 above)
Comment (DongyunRobotis) [reply physical_ai_server/physical_ai_server/physical_ai_server.py]: Good

File: physical_ai_server/physical_ai_server/physical_ai_server.py (hunk 5 above)
Comment (DongyunRobotis) [reply physical_ai_server/physical_ai_server/physical_ai_server.py]: Good

**Review by DongyunRobotis — APPROVED** [review]
LGTM
